### PR TITLE
Data prefetching and multi-GPU

### PR DESCRIPTION
**Pull request by seed93**
Notice:
We didn't sort input data according to its seqLength and use a fixed batchsize. We need an indicator to tell LSTM the seqLengths.

### CODE REVIEW

**Comment by SeanNaren** [comment]
This is super awesome thanks a bunch! I'll run through the code as soon as I can.

File: DeepSpeechModel.lua [hunk 1]
```diff
@@ -1,32 +1,44 @@
...
 
-    local model = nn.Sequential()
-    model:add(cudnn.SpatialConvolution(1, 32, 41, 11, 2, 2))
-    model:add(cudnn.SpatialBatchNormalization(32))
-    model:add(cudnn.ReLU(true))
-    model:add(cudnn.SpatialConvolution(32, 32, 21, 11, 2, 1))
-    model:add(cudnn.SpatialBatchNormalization(32))
-    model:add(cudnn.ReLU(true))
-    model:add(cudnn.SpatialMaxPooling(2, 2, 2, 2))
+    GRU = false
```
Comment (SeanNaren) [inline DeepSpeechModel.lua]: is it supposed to say false here? I assume this was done for testing

**Comment by SeanNaren** [comment]
Just a heads up, I'll pull this branch into master once we've fixed the issue with variable sequence lengths, add batch norm to LSTMs and update README with new instructions and improvements.

@seed93 and his team have done a great job so far, props to them :)

**Comment by SeanNaren** [comment]
You guys are working ridiculously fast, can't thank you enough! To make your lives a bit easier I'll put the modified Criterion structure into the main lua--nnx library.

File: util.lua [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+        -- make label
+        local label, modified_trans = trans2tokens(trans, mapper)
+
+        -- make spect
+        local wave = audio.load(root_path..wave_path)
+        local spect = audio.spectrogram(wave, windowSize, 'hamming', stride) -- freq-by-frames tensor
+
+        -- put into lmdb
+        txn_spect:put(cnt, spect:byte())
```
Comment (SeanNaren) [inline util.lua:104]: Removing the :byte() conversion still works here. It seems to clip the decibels otherwise?

File: prepare_an4/prepare.sh [hunk 3]
```diff
@@ -0,0 +1,19 @@
+#!/bin/sh
```
Comment (SeanNaren) [inline prepare_an4/prepare.sh:1]: Nice job with the file, I've created a prepare an4 script that does everything, from downloading to extracting. Code here is [here](https://gist.github.com/SeanNaren/20608497249016ff645b942809d69f99)

File: loader.lua [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+function loader:nxt_batch(inds, flag)
+    --[[
+        return a batch by loading from lmdb just-in-time
+
+        flag: indicates whether to load trans
+
+        TODO we allocate 2 * batch_size space
+    --]]
+    local tensor_list = {}
```
Comment (SeanNaren) [inline loader.lua]: I've had complaints of memory issues with storing it like this due to LuaJIT 2gb limit. What works is to use [this](https://github.com/torch/tds) and change this to local tensor_list = tds.Vec(). Also you have to import above tds = require 'tds'.

Also you would have to change table.insert(tensor_list, tensor) to tensor_list:insert(tensor). Everything else is good!

File: Network.lua [hunk 5]
```diff
@@ -129,14 +206,14 @@ end
...
+    saveDataParallel(saveName, self.model)
 end
 
+
 --Loads the model into Network.
 function Network:loadNetwork(saveName)
-    local model = torch.load(saveName)
-    self.model = model
+    self.model = loadDataParallel(saveName, self.nGPU)
     model:evaluate()
```
Comment (SeanNaren) [inline Network.lua]: I think you may need to change this to self.model:evaluate()

**Comment by seed93** [comment]
We finished mask for variant seqLengths support. However, multigpu converges much slower than single gpu. We are still working on it. Thank you!

**Comment by SeanNaren** [comment]
@seed93 have a look at the criterion I've opened up a PR for [here](https://github.com/clementfarabet/lua---nnx/pull/61). It has batchFirst support which may be of use to you (I've had reports from @willfrey that this was an issue). 

I've added a few notes as to small bugs that are not very big but may help you guys out. Also awesome stuff with the mask support!

Finally any chance you might be able to return the WER calculations to what they were initially? I understand it has changed a bit and maybe after the PR I could work on reverting it similar to what it was (showing all the different WER, and showing the WER on each epoch).

File: Network.lua [hunk 6]
```diff
@@ -4,98 +4,164 @@ require 'BRNN'
...
     end
 
+    -- ==========================================================
+    -- training
+    -- ==========================================================
     local currentLoss
     local startTime = os.time()
-    local dataSetSize = dataset:size()
+    -- local dataSetSize = self.indexer.len_num -- obtained when calling prep_same_len_inds
+    local dataSetSize = 48
```
Comment (SeanNaren) [inline Network.lua]: Would the below work? This involved me having to create a batch_size variable in the init using the networkParams.

`local dataSetSize = math.ceil(self.indexer.lmdb_size / self.batch_size)`

Comment (seed93) [inline Network.lua]: I will merge your code the day after tomorrow, sorry for that.

File: loader.lua (hunk 4 above)
Comment (seed93) [inline loader.lua]: Sounds much more compatible! Do we need 2gb data in a minibatch?

File: Network.lua (hunk 5 above)
Comment (seed93) [inline Network.lua]: This bug has been fixed, sorry for that.

File: loader.lua (hunk 4 above)
Comment (SeanNaren) [inline loader.lua]: It was a bit strange but hopefully this fix will at-least be long term if someone decides to really up the batch size!

**Comment by TangXing** [comment]
@SeanNaren hi when i run " th AN4CTCTest.lua" i get this error!
torch/install/share/lua/5.1/cudnn/RNN.lua:281: Assertion `THCudaTensor_checkGPU(state, 1, self_)' failed.  at /tmp/luarocks_cutorch-scm-1-6942/cutorch/lib/THC/THCTensorMath.cu:31
stack traceback:
    [C]: in function 'zero'
    ...//torch/install/share/lua/5.1/cudnn/RNN.lua:281: in function 'updateOutput'
    ./MaskRNN.lua:29: in function 'func'
    .../torch/install/share/lua/5.1/nngraph/gmodule.lua:311: in function 'neteval'
    .../torch/install/share/lua/5.1/nngraph/gmodule.lua:346: in function 'forward'
    ./wer_tester.lua:83: in function 'get_wer'
    ./Network.lua:80: in function 'testNetwork'
    AN4CTCTest.lua:30: in main chunk
    [C]: in function 'dofile'
    .../torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
    [C]: at 0x004064d0

**Comment by seed93** [comment]
I will check this tomorrow, thank you

**Comment by SeanNaren** [comment]
I'll also have a look when I start merging this. I plan to start the merge tomorrow as long as @seed93 thinks its appropriate. Lots of refactoring/updating of docs to do so might take me some time :)

**Comment by SeanNaren** [comment]
Right fully merged into the master branch, lots of changes on top of what you guys have done that I would love feedback on. Also multi-GPU still is broken, let me know if you can take a stab at it, else I can take a look. Great work as usual :)
